[quant][graphmode][fx][test] Add per op test for graph mode quant on fx

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43335 [quant][graphmode][fx] Add support for batchnorm relu
* #43334 [quant][graphmode][fx] Add support for batchnorm
* #43333 [quant][graphmode][fx] Add support for cat
* #43332 [quant][graphmode][fx] Add support for add relu
* #43331 [quant][graphmode][fx] Add support for add
* #43287 [quant][graphmode][fx] Add support for conv module + relu
* #43286 [quant] Make OP_LIST_TO_FUSER_METHOD public
* #43285 [quant][graphmode][fx] Add support for conv module
* **#43229 [quant][graphmode][fx][test] Add per op test for graph mode quant on fx**
* #43175 [quant][graphmode][fx] Add graph mode quantization on fx

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

jerryzh168 committed:
[quant][graphmode][fx][test] Add per op test for graph mode quant on fx

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

supriyar commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -358,6 +367,65 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
+        result = qgraph(*inputs)
+        result_debug = qgraph_debug(*inputs)
+
+        assert (result - result_debug).abs().max() == 0, \
```
use `self.assertEqual`?

supriyar commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -358,6 +367,65 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
+                return
+
+        self.assertTrue(False, 'node:' + str(target_node) +
+                        ' not found in the graph module')
```
nit: move to previous line?

supriyar commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -358,6 +367,65 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
+        """ Check if GraphModule contains the target node
+        Args:
+            graph_module: the GraphModule instance we want to check
+            target_op: a tuple of 2 elements, first element is
```
s/target_op/target_node

supriyar commented on test/quantization/test_quantize_fx.py:
```diff
@@ -102,15 +112,74 @@ def forward(self, x):
+
+        data = [torch.rand((1, 30), dtype=torch.float)]
+        options = itertools.product(
+            [(ModuleLinear(has_relu=False), True),
+             (FuncLinear(has_relu=False), False)],
```
are we missing a check for `has_relu=True`?

jerryzh168 replied:
we do have it in L181, although some of the functionalities(functional linear + relu) are not supported yet

supriyar replied:
I see, I think we also need to add linear dynamic + relu fusion as well.

supriyar reviewed (approved):
lgtm, left a few comments

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -565,6 +566,8 @@ def prepare_dynamic(self, model, qconfig_dict, inplace=False):
     def convert(self, observed, inplace=False, debug=False):
         assert self.activation_post_process_map is not None
+        # move to cpu since we only have quantized cpu kernels
+        observed.eval().cpu()
```
thoughts on if we can let the user do this instead, and assert that the model is on CPU (or handle GPU)?  I think doing it here might be surprising to users

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
+        """ Check if GraphModule contains the target node
+        Args:
+            graph_module: the GraphModule instance we want to check
+            target_node: a tuple of 2 elements, first element is
```
thanks for the docblock! can we provide an example call as well?

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
+        self.assertTrue(False, 'node:' + str(target_node) +
+                        ' not found in the graph module')
+
+    def printGraphModule(self, graph_module):
```
not feedback for this PR, more just wondering if we can move this to GraphModule core at some point?

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
+                node_info += ' module type:' + repr(type(modules[n.target]))
+            print(node_info)
+
+    def checkGraphModeFxOp(self, model, inputs, quantized_node, quant_type=QuantType.STATIC, debug=False):
```
optional: maybe the name could be `quantWithFxAndCheckQuantNode` or something like that?

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
+        if quant_type == QuantType.DYNAMIC:
+            prepared = quantizer.prepare_dynamic(fused, qconfig_dict)
+        else:
+            prepared = quantizer.prepare(fused, qconfig_dict)
```
not feedback for this PR, more a question, should we have `prepare_qat` to be consistent with eager mode?

jerryzh168 committed:
Update on "[quant][graphmode][fx][test] Add per op test for graph mode quant on fx"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

[ghstack-poisoned]
